src/cpp/session/modules/SessionFiles.cpp

```diff
@@ -660,41 +664,68 @@ Error completeUpload(const core::json::JsonRpcRequest& request,
+                              const std::string& originalFilename,
+                              json::Array* pOverwritesJson,
+                              bool& rUnzipfound)
 {
-   // query for all of the paths in the zip file
-   core::system::ProcessResult result;
-   Error error = core::system::runCommand("unzip -Z1 " + uploadedZipFile.getAbsolutePath(),
-                                          core::system::ProcessOptions(),
-                                          &result);
-   if (error)
-      return error;
+   // unable to use R's unzip here in worker thread, using system's unzip instead
+   // try a couple locations for unzip, in case it's not on user's PATH
+   std::vector<std::string> pathsToTry({
+                                       "unzip",
+                                       "/usr/bin/unzip"});
+   for (auto &path : pathsToTry) {
+      core::system::ProcessResult result;
+      Error error = core::system::runCommand(path + " -Z1 " + uploadedZipFile.getAbsolutePath(),
+                                             core::system::ProcessOptions(),
+                                             &result);
 
-   if (result.exitStatus == 0)
-   {
-      std::vector<std::string> zipFileListing;
-      boost::split(zipFileListing, result.stdOut, boost::is_any_of("\n"));
-
-      // check for overwrites
-      for (std::vector<std::string>::const_iterator
-           it = zipFileListing.begin();
-           it != zipFileListing.end();
-           ++it)
+      if (error)
+         return error;
+
+      // If any of the paths were valid, check for overwrites and return
+      if (result.exitStatus == 0)
       {
-         // don't count empty lines
-         if ((*it).empty()) continue;
+         rUnzipfound = true;
+
+         std::vector<std::string> zipFileListing;
+         boost::split(zipFileListing, result.stdOut, boost::is_any_of("\n"));
+
+         // check for overwrites
+         for (std::vector<std::string>::const_iterator
+              it = zipFileListing.begin();
+              it != zipFileListing.end();
+              ++it)
+         {
+            // don't count empty lines
+            if ((*it).empty()) continue;
+
+            FilePath filePath = destDir.completePath(*it);
+            if (filePath.exists())
+               pOverwritesJson->push_back(module_context::createFileSystemItem(filePath));
+         }
 
-         FilePath filePath = destDir.completePath(*it);
-         if (filePath.exists())
-            pOverwritesJson->push_back(module_context::createFileSystemItem(filePath));
+         return Success();
+      }
+      else if (result.exitStatus == 127)
+      {
+         // bash return code 127 - "command not found"
+         // try another unzip location
+         continue;
+      }
+      else
+      {
+         return systemError(result.exitStatus,
+                            "Unexpected result for unzip command",
+                            ERROR_LOCATION);
```

Comment:
I don't think we want to supply `result.exitStatus` to `systemError` here. I know we do it in general, but I think it leads to some confusing error messages because it ends up being mapped to an `errno` value. Exit codes from processes have a different meaning than `errno`s of the same value. For example, exit code 127 (as you used in your code above) indicates that the executable could not be found. `errno` 127 means "Key has expired" (`EKEYEXPIRED`). 

I'm not sure what we should pass in place, but I think we should include the exit status in the message instead of as the error code. @kevinushey Do you have any ideas of what would be a better error type? Maybe we need to make a new class of errors for process exits.

Reply:
In this case, the exit code of 127 is handled explicitly above, so I think this is okay. (The other alternative would be to test for existence of the program; e.g. using `core::system::findProgramOnPath()` and then consider reporting a FileNotFoundError or something similar if it, or any hard-coded hint locations, cannot be resolved)

Reply:
I was just using 127 as an example as it has a nice specific meaning as an exit code. An exit code of 1 just means the process exited abnormally, but `errno` 1 is `EPERM` (operation not permitted). Exit odes between 1 and 127 (or 2 and 126, depending on your opinion) have whatever meaning the application developer chooses. Exit codes `128 + n` indicate the process exited with the signal `n`. There are no exit codes that line up with `errno` (at least not on purpose in the general case).

Reply:
Ah, sorry, you're exactly right. I guess we just need a more generic "unzip returned error code %s" error / message here.

Reply:
Would it be more clear if we used some "dummy/close enough" system error code, and then including the actual return code/error messages in the message?

``` suggestion
         // Not really a good error code to indicate that our bash command had an error
         // *nix:  1 = EPERM - Operation not permitted
         // WIN32: 1 = ERROR_INVALID_FUNCTION - Invalid function
         return systemError(1,
                            "Unexpected result for unzip command: " + 
                            std::to_string(result.exitStatus) + 
                            " - " + 
                            result.stdOut +
                            ": " +
                            result.stdErr,
                            ERROR_LOCATION);
```

Reply:
This might be too big for your change, but I think we'd be better off creating a different category of errors like "ProcessExitCode" errors or something. In the meantime, I think you should change `systemError` to `unknownError` so that at least we don't provide a message that looks like it has meaning (e.g. `"system error 1 (Operation not Permitted): <message here>"`) when really it doesn't. At least `unknownError` won't give false information.

Reply:
Ah! Thank you! I hadn't noticed `unknownError` was an option. Using that for now.